File: quicksort.cpp
#include <iostream>
using namespace std;

void printArray(int arr[], int size)
{
    for (int i = 0; i < size; i++)
        cout << arr[i] << " ";
    cout << endl;
}
void swap(int *a, int *b)
{
    int temp = *a;
    *a = *b;
    *b = temp;
}

int partition(int arr[], int low, int high)
{
    int pivot = arr[high]; // pivot
    int i = (low - 1);     // Index of smaller element
    for (int j = low; j <= high - 1; j++)
    {
        if (arr[j] < pivot)
        {
            i++; // Increment index of smaller element
            swap(&arr[i], &arr[j]);
        }
    }
    swap(&arr[i + 1], &arr[high]);
    printArray(arr, sizeof(arr) / sizeof(arr[0]));

    return (i + 1);
}

void quickSort(int arr[], int low, int high)
{
    if (low < high)
    {
        int pi = partition(arr, low, high); // Partition index

        quickSort(arr, low, pi - 1);  // Sort elements before partition
        quickSort(arr, pi + 1, high); // Sort elements after partition
    }
}

int main()
{
    int arr[] = {10, 7, 8, 9, 1, 5};
    int n = sizeof(arr) / sizeof(arr[0]);
    quickSort(arr, 0, n - 1);
    cout << "Sorted array: ";
    printArray(arr, n);
    return 0;
}

File: js/quickSort.js
function quickSort(arr) {
  // Base case: If the array is of length 0 or 1, it's already sorted
  if (arr.length <= 1) {
    return arr;
  }

  // Helper function to partition the array
  function partition(arr, low, high) {
    let pivot = arr[high]; // Choose the last element as the pivot
    let i = low - 1; // Pointer for the smaller element

    // Traverse the array from low to high-1
    for (let j = low; j < high; j++) {
      // If current element is less than or equal to the pivot
      if (arr[j] <= pivot) {
        i++; // Move the pointer for the smaller element
        [arr[i], arr[j]] = [arr[j], arr[i]]; // Swap elements
      }
    }

    // Swap the pivot element with the element at i+1 to place the pivot in the correct position
    [arr[i + 1], arr[high]] = [arr[high], arr[i + 1]];
    return i + 1; // Return the index of the pivot
  }

  // Recursive function to sort the subarrays
  function quickSortRecursive(arr, low, high) {
    if (low < high) {
      let pi = partition(arr, low, high); // Partitioning index
      quickSortRecursive(arr, low, pi - 1); // Sort left subarray
      quickSortRecursive(arr, pi + 1, high); // Sort right subarray
    }
  }

  // Start the sorting process
  let arrCopy = [...arr]; // Make a copy of the array to avoid mutation of the original
  quickSortRecursive(arrCopy, 0, arrCopy.length - 1);
  return arrCopy;
}

// Example usage:
const arr = [10, 7, 8, 9, 1, 5];
console.log(quickSort(arr)); // Output: [1, 5, 7, 8, 9, 10]
